libs/wire-api/src/Wire/API/MLS/Serialisation.hs

```diff
@@ -81,6 +82,19 @@ parseMLSVector getItem = do
       pos <- bytesRead
       (:) x <$> (if pos < endPos then go endPos else pure [])
 
+serialiseMLSVector ::
+  forall w a.
+  (Binary w, Integral w) =>
+  (a -> Put) ->
+  [a] ->
+  Put
+serialiseMLSVector p as = do
+  -- every element is long the same number of bytes when serialised. If there is
+  -- no single element in the vector, the whole length is 0 anyways.
+  let elLen = maybe 0 (LBS.length . runPut . p) . listToMaybe $ as
+  put @w . fromIntegral $ elLen * fromIntegral (length as)
+  traverse_ p as
```

Comment:
This doesn't look right to me. Why is every element the same length?

To be fair, it's not easy to do this efficiently. As a compromise, we can do the following:
 - serialise the whole vector ignoring the size
 - convert to lazy bytestring
 - put count
 - put bytestring

To get good performance we need imperative access to the buffer, which is probably possible, but requires the unsafe API. Maybe the above is good enough.

Reply:
Good points! I obviously overlooked at this.

I changed the implementation and round-trip tests for `serialiseMLSVector` pass. I am not too familiar with how efficient the conversion from a lazy bytestring to a strict bytestring is, but that's what I did.

Reply:
Lazy bytestrings are essentially lists of strict bytestrings (represent their concatenation). Converting from lazy to strict amounts to actually performing the concatenation (i.e. basically allocating a big buffer, copying, and likely leaving the original bytestrings to the GC). So it's basically a bunch of extra copies and allocations, good to avoid if possible.